src/lib/operators/table_scan.hpp

```diff
@@ -25,7 +28,165 @@ class TableScan : public AbstractOperator {
+      auto table = segment->referenced_table();
+
+      const Chunk& chunk = table->get_chunk(chunk_id);
+      const auto actual_segment = chunk.get_segment(column_id);
+
+      const std::string column_type = table->column_type(_column_id);
+
+      resolve_data_type(column_type, [&](auto type) {
+        using Type = typename decltype(type)::type;
+        const Type search_value = type_cast<Type>(_search_value);
+
+        // handle value segment
+        const auto value_segment = std::dynamic_pointer_cast<ValueSegment<Type>>(actual_segment);
+        if (value_segment) {
+          if (_evaluate_scan_predicate<Type>(search_value, value_segment->values().at(real_chunk_offset))) {
+            matched_row_ids->push_back(RowID{chunk_id, real_chunk_offset});
+          }
+          return;
+        }
+
+        // handle dictionary segment
+        const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<Type>>(actual_segment);
+        if (dictionary_segment) {
+          if (_evaluate_scan_predicate<Type>(search_value, dictionary_segment->get(real_chunk_offset))) {
+            matched_row_ids->push_back(RowID{chunk_id, real_chunk_offset});
+          }
+          return;
+        }
+        throw std::logic_error("Unknown segment type");
+      });
+    }
+  }
+
+  template <typename T>
+  void _collect_matched_rows_for_value_segment(const ChunkID chunk_id, const std::shared_ptr<ValueSegment<T>>& segment,
+                                               const std::shared_ptr<PosList>& matched_row_ids) const {
+    const T search_value = type_cast<T>(_search_value);
+    ChunkOffset chunk_offset{0};
+    for (const T& value : segment->values()) {
+      if (_evaluate_scan_predicate(search_value, value)) {
+        matched_row_ids->push_back(RowID{chunk_id, chunk_offset});
+      }
+      chunk_offset++;
+    }
+  }
+
+  template <typename T>
+  void _collect_matched_rows_for_dictionary_segment(const ChunkID chunk_id,
+                                                    const std::shared_ptr<DictionarySegment<T>>& segment,
+                                                    const std::shared_ptr<PosList>& matched_row_ids) const {
+    const ValueID lower_bound_value = segment->lower_bound(_search_value);
+    const T search_value = type_cast<T>(_search_value);
+
+    const bool value_exists =
+        (lower_bound_value != INVALID_VALUE_ID) && (segment->value_by_value_id(lower_bound_value) == search_value);
+
+    const auto attribute_vector = segment->attribute_vector();
+    const auto attribute_vector_size = attribute_vector->size();
+
+    ValueID value_id;
+    for (size_t chunk_offset = 0; chunk_offset < attribute_vector_size; ++chunk_offset) {
+      value_id = attribute_vector->get(chunk_offset);
+      if (_evaluate_scan_predicate_for_dictionary_segment(value_id, lower_bound_value, value_exists)) {
+        matched_row_ids->push_back(RowID{chunk_id, static_cast<ChunkOffset>(chunk_offset)});
```

Comment:
`emplace_back` could be faster here:
```suggestion
        matched_row_ids->emplace_back(RowID{chunk_id, static_cast<ChunkOffset>(chunk_offset)});
```